File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.py
import sys
from collections import deque
input = sys.stdin.readline

t = int(input())
for _ in range(t):
    n = int(input())
    x, y = map(int, input().split())
    stores = []
    for i in range(n):
        a, b = map(int, input().split())
        dist = abs(a-x) + abs(b-y)
        stores.append((a, b, dist))
    festival = list(map(int, input().split()))
    stores = sorted(stores, key=lambda x: x[2])
    visited = [False]*n
    answer = 0
    q = deque([(x, y)])

    while q:
        nx, ny = q.popleft()
        if abs(festival[0] - nx) + abs(festival[1] - ny) <= 1000:
            answer = 1
            break

        for i in range(n):
            if not visited[i]:
                dist = abs(stores[i][0] - nx) + abs(stores[i][1] - ny)

                if dist <= 1000:
                    q.append((stores[i][0], stores[i][1]))
                    visited[i] = True

    if answer:
        print('happy')
    else:
        print('sad')

File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.cc
#include <bits/stdc++.h>
using namespace std;
#define FIO                                                                    \
  ios_base::sync_with_stdio(false);                                            \
  cin.tie(NULL);                                                               \
  cout.tie(NULL);
typedef long long ll;

int t, n, sx, sy, ex, ey;
pair<int, int> store[100];
int main() {
  FIO;
  cin >> t;
  while (t--) {
    bool visited[100] = {false};
    string answer = "sad";
    deque<pair<int, int>> dq;

    cin >> n >> sx >> sy;
    for (int i = 0; i < n; ++i) {
      cin >> store[i].first >> store[i].second;
    }
    cin >> ex >> ey;

    dq.push_back(make_pair(sx, sy));
    while (!dq.empty()) {
      int x = dq.front().first, y = dq.front().second;
      dq.pop_front();

      if (abs(ex - x) + abs(ey - y) <= 1000) {
        answer = "happy";
        break;
      }

      for (int i = 0; i < n; ++i) {
        if (!visited[i] &&
            abs(store[i].first - x) + abs(store[i].second - y) <= 1000) {
          visited[i] = true;
          dq.push_back(store[i]);
        }
      }
    }
    cout << answer;
    if (t)
      cout << "\n";
  }
}